pkg/version/version.go

```diff
@@ -191,3 +197,81 @@ func IsReleaseVersion() bool {
 		GetMainVersion() != "" &&
 		!strings.Contains(GetMainVersion(), "-")
 }
+
+// IsPriorToScannerV4 returns true if version represents a version of ACS from prior to the
+// introduction of Scanner V4. Will return an error if cannot determine result.
+func IsPriorToScannerV4(version string) (bool, error) {
+	var err error
+	parsed, err := parseVersion(version)
+	if err != nil {
+		return false, err
+	}
+
+	x := parsed.MarketingMajor
+	y := parsed.EngRelease
+	z := parsed.PatchLevel
+
+	if x < 4 || y < 3 || (y == 3 && (z == "" || z != "x")) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Variants breaks a version into a series of version strings starting with
+// the most specific to the least specific, stopping at X.Y. It uses
+// parseVersion to validate the format.
+func Variants(version string) ([]string, error) {
+	_, err := parseVersion(version)
+	if err != nil {
+		return nil, err
+	}
+
+	res := []string{version}
+
+	i := strings.LastIndex(version, "-")
+	for i != -1 {
+		res = append(res, version[:i])
+		version = version[:i]
+		i = strings.LastIndex(version, "-")
+	}
+
+	i = strings.LastIndex(version, ".")
+	for i != -1 && strings.Count(version, ".") > 1 {
+		res = append(res, version[:i])
+		version = version[:i]
+		i = strings.LastIndex(version, ".")
+	}
+
+	return res, nil
+}
+
+// parseVersion mimics parseMainVersion but allows for versions to be in a
+// format that would otherwise not be a valid main version (such as X.Y).
+func parseVersion(version string) (parsedMainVersion, error) {
+	parsed, err := parseMainVersion(version)
+	if err != nil && !errors.Is(err, ErrInvalidNumberOfComponents) {
+		return parsedMainVersion{}, err
+	}
+
+	if err == nil {
+		return parsed, nil
+	}
+
+	before, _, _ := strings.Cut(version, "-")
+	parts := strings.Split(before, ".")
+	if len(parts) != 2 {
+		return parsedMainVersion{}, fmt.Errorf("%w (expected 2-4, got %d)", ErrInvalidNumberOfComponents, len(parts))
```

Comment:
why 4? oh for the really old versioning scheme?

Reply:
Yes it is, since `parseMainVersion` error mentions it - figured pass it along:

https://github.com/stackrox/stackrox/blob/219cb5d6e39bc72fad4fb874a249dbd4f92fcad3/pkg/version/version.go#L95